dlib/dnn/trainer.h

```diff
@@ -1194,6 +1194,9 @@ namespace dlib
             job.have_data.resize(devs);
             job.test_only = test_only;
 
+            // check if the iterator points to anything
+            const bool have_labels = sizeof(*lbegin) != 1;
```

Comment:
Oof, yeah, I see the bug you are trying to fix.  This bit of code here is not right though.  You can't dereference an iterator to nothing, that's undefined behavior.  And it might not refer to a type that has a size of 1 as well.  So something like `std::optional<label_iterator>` is needed instead to indicate if we actually have a `lbegin`.  Can't use `std::optional` in dlib yet though :(

Maybe use `std::pair<bool, label_iterator>` instead.

Reply:
Yes, I wasn't convinced about this either. I will revert the change.
I thought checking the `labels[i].size()` would work (my first attempt). But it didn't, so i duplicated the `send_job`code...

I'm not sure i get what you're saying about the new iterator type, I'll check later. Thank you!